node, graph, store, chain: Remove unneeded blocks from the block cache

lutter opened the pull request:
Fixes https://github.com/graphprotocol/support/issues/45

Jannis commented on chain/ethereum/src/block_ingestor.rs:
```diff
@@ -73,40 +73,69 @@ where
+                        // Continue polling even if polling failed
+                        future::ok(())
+                    })
+                    .then(move |result| {
```
You could use `inspect` here, then you don't need to pass `result` through:
```rust
.inspect(move |_| {
  ...
})
```
I do wonder though, wouldn't this better be implemented via a separate Tokio task / thread?

lutter replied:
What's the concern here? When you do cleanups, the overall slowdown of the system because you need to continually download blocks will far outweigh any slowdown from adding a `delete` query into the blockstream.

Jannis replied:
I'm not sure I understand the "continually download blocks" part. I was just wondering if running this delete query every time there is a new chain head would slow down the ingestion, but since we can configure how frequently that happens, I think we're good here. It was more a question around a separation of concerns.

Jannis commented on node/src/main.rs:
```diff
@@ -51,6 +51,13 @@ lazy_static! {
+        .map(|s| u64::from_str(&s)
+             .unwrap_or_else(|_| panic!("failed to parse env var GRAPH_BLOCK_CACHE_CLEANUP_FREQ")))
+        .map(|secs| Duration::from_secs(secs))
+        .unwrap_or(Duration::from_secs(0));
```
Can we move this into `BlockIngestor`, rename it to `GRAPH_ETHEREUM_BLOCK_CLEANUP_INTERVAL`, and also drop the frequency argument to `into_polling_stream()`? I feel this is internal to Ethereum, for other chains in the future we may want to configure different intervals.

lutter replied:
I am actually thinking that it shouldn't be an interval at all, and we should do that cleanup after each block we ingest, i.e., make that environment variable a boolean. That way, we will most often clean up just a small number of blocks, and it'll help in keeping `ethereum_blocks` small.

lutter replied:
I made it a boolean, and changed the name to `GRAPH_ETHEREUM_CLEANUP_BLOCKS` and moved it into the block ingestor

Jannis commented on store/postgres/src/store.rs:
```diff
@@ -1268,6 +1268,50 @@ impl ChainStore for Store {
+        // subgraph's head block, but retain the genesis block. We stay
+        // behind the slowest subgraph so that we do not interfere with its
+        // syncing activity.
+        // Only consider active subgraphs that have not failed
```
That's pretty cool!

Jannis commented on store/postgres/src/store.rs:
```diff
@@ -1268,6 +1268,50 @@ impl ChainStore for Store {
+                            subgraphs.entities ds
+                      where d.entity='SubgraphDeployment'
+                        and a.entity='SubgraphDeploymentAssignment'
+                        and ds.entity = 'EthereumContractDataSource'
```
Nit: The spacing around `=` is inconsistent.

leoyvens commented:
Exactly how it all works is not clear in my mind right now, but the block ingestor makes sure to keep the latest `REORG_THRESHOLD` blocks in the database, and I believe the block stream relies on that assumption. Could deleting these recent blocks be a potential issue?

lutter commented:
> Exactly how it all works is not clear in my mind right now, but the block ingestor makes sure to keep the latest `REORG_THRESHOLD` blocks in the database, and I believe the block stream relies on that assumption. Could deleting these recent blocks be a potential issue?

Good point. I think the minimum block number we keep needs to be `min(subgraph heads, chain head - REORG_THRESHOLD)`

lutter commented:
Addressed all review comments. We now only clean up blocks that are before the minimum of any subgraph heads and before `ANCESTOR_COUNT` of the chain head.

Jannis commented on chain/ethereum/src/block_ingestor.rs:
```diff
@@ -84,29 +92,57 @@ where
+                            match static_self
+                                .chain_store
+                                .cleanup_cached_blocks(static_self.ancestor_count)
+                            {
```
I liked the ability to control the interval. Do we think this is cheap enough to run every time?

Jannis replied:
Ah, well, it's a development time thing, so let's not worry about it. 😄

lutter replied:
That, and I really wouldn't know what to tell users how to choose a good interval. Testing this locally with one syncing subgraph, it would generally remove one or two blocks, and that did not slow the block stream down at all, it's very quick. I had (from previous syncs) ~ 20k blocks in the DB; cleanup whittled it down to about 2k blocks while it was syncing without causing any noticeable delay.

On the 'continually download blocks': yeah, that phrasing could be improved, I just wanted to say something sufficiently scary so people don't turn this on in production settings. But I am open to rewording that, just not sure how.

lutter commented:
Rebased to latest master
